astropy/io/registry/core.py

```diff
@@ -0,0 +1,411 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+import os
+import sys
+from collections import OrderedDict
+
+import numpy as np
+
+from .base import IORegistryError, _UnifiedIORegistryBase
+
+__all__ = ['UnifiedIORegistry', 'UnifiedInputRegistry', 'UnifiedOutputRegistry']
+
+
+PATH_TYPES = (str, os.PathLike)  # TODO! include bytes
+
+
+# -----------------------------------------------------------------------------
+
+class UnifiedInputRegistry(_UnifiedIORegistryBase):
+    """Read-only Unified Registry.
+
+    .. versionadded:: 5.0
+
+    Examples
+    --------
+    First let's start by creating a read-only registry.
+
+    .. code-block:: python
+
+        >>> from astropy.io.registry import UnifiedInputRegistry
+        >>> read_reg = UnifiedInputRegistry()
+
+    There is nothing in this registry. Let's make a reader for the
+    :class:`~astropy.table.Table` class::
+
+        from astropy.table import Table
+
+        def my_table_reader(filename, some_option=1):
+            # Read in the table by any means necessary
+            return table  # should be an instance of Table
+
+    Such a function can then be registered with the I/O registry::
+
+        read_reg.register_reader('my-table-format', Table, my_table_reader)
+
+    Note that we CANNOT then read in a table with::
+
+        d = Table.read('my_table_file.mtf', format='my-table-format')
+
+    Why? because ``Table.read`` uses Astropy's default global registry and this
+    is a separate registry.
+    Instead we can read by the read method on the registry::
+
+        d = read_reg.read(Table, 'my_table_file.mtf', format='my-table-format')
+
+    """
+
+    def __init__(self):
+        super().__init__()  # set _identifiers
+        self._readers = OrderedDict()
+        self._registries["read"] = dict(attr="_readers", column="Read")
+        self._registries_order = ("read", "identify")
+
+    def get_formats(self, data_class=None, filter_on="Read"):
```

Comment:
I think the ``get_formats`` in ``UnifiedIORegistry`` is necessary (unless we want to change the API by updating the kwarg ``readwrite`` to ``filter_on``) but I can certainly get rid of the overrides in ``UnifiedInputRegistry`` and ``UnifiedOutputRegistry``. The default behavior is to return a combination of all registries, so that part is taken care of. Thanks!